sequential_handler: simplify to use less goroutines

jhenstridge opened the pull request:
This is an attempt to reduce the number of goroutines used by the SequentialSignalHandler implementation, which seemed to occasionally trip up the race detector in the tests.  The main changes are:

1. spin up one goroutine to handle each signal channel rather than one per signal per channel.
2. move the handling of the queue of undelivered signals into this goroutine
3. remove the locks and wait group that were serialising access to the queue, since it is now private to a single goroutine.

The third `done` channel is needed to ensure the `bufferSignals` goroutine has completed.  Without it, it's possible that the goroutine will still be stuck in the `select` call when the user closes the signal channel, leading to a panic.

I've kept the use of the list type, but it may be more effective to use a slice if we expect the number of buffered signals to be small.

jsouthworth commented:
This looks much clearer to me.

@patrickmeiring since you are the original author of this function do you have any comments? Any thoughts on whether we should use a list or a slice as @jhenstridge questions above?

patrickmeiring commented:
Looks like a useful improvement to me, and the new implementation respects the requirement to avoid blocking the main DBus message processing routine. 

My only suggestion would be to preserve the documentation somewhere (e.g. by adding a method comment) that this is in fact a requirement of the deliver method, so that we can check future refactorings also satisfy this requirement.

From memory, I kept the list because I thought it best modeled a queue ADT. If we're switching for performance reasons, it would be good to have some benchmarks to back this up. I vaguely recall testing a different SequentialHandler implementation at the time I submitted the original PR (somewhere along the lines of this CL, using just one goroutine) and finding it had a negligible impact on overall throughput (IIRC, because signal delivery is but a small part of the overall task of receiving/parsing/dispatching signals). As it seemed slightly more complex to do everything on one goroutine, I didn't submit it. I didn't consider that the extra goroutines would cause trouble with race detectors, though.

In the absence of benchmarks, I would just go with whatever you think most cleanly expresses the intent.

jhenstridge commented:
I've converted over to using a slice, which simplified the code a bit, and added a version of the comment from `deferredDeliver` to `bufferSignals`.

It probably is worth doing some benchmarks: I imagine this one would do better when the code consuming the signals f alls behind.  But it also keeps a goroutine around at all times to manage the queue.  Maybe a middle ground would be to attempt direct delivery of the signals up until delivery would block, and only them spin up the goroutine?

If the performance and memory usage is fine, I wonder if it would be worth just making this the default SignalHandler implementation too.

jsouthworth commented:
@jhenstridge are you working on the follow ups independently of this PR or are they blocking the PR? Let me know if you are ready to merge. Thanks.

jhenstridge commented:
Sorry for not updating.  I don't know if I'll have time to add that extra optimisation.  I believe that what is here right now is an improvement over master.

I'm not sure that full logic to start and stop the goroutine on demand is worth the trouble.  If might make sense to delay starting the goroutine until we see the user provided channel reports full.  With that change, it would probably be worth making this handler the default.

jsouthworth commented:
Got it, thanks. I'll merge this as is and the additional work can be done separately.
